casbin/distributed_enforcer.py

```diff
@@ -0,0 +1,132 @@
+from casbin import SyncedEnforcer
+import logging
+
+from casbin.persist import batch_adapter
+from casbin.model.policy_op import PolicyOp
+from casbin.persist.adapters import update_adapter
+
+
+class DistributedEnforcer(SyncedEnforcer):
+    """DistributedEnforcer wraps SyncedEnforcer for dispatcher."""
+
+    def __init__(self, model=None, adapter=None):
+        self.logger = logging.getLogger()
+        SyncedEnforcer.__init__(self, model, adapter)
+
+    def add_policy_self(self, should_persist, sec, ptype, rules):
+        """
+        AddPolicySelf provides a method for dispatcher to add authorization rules to the current policy.
+        The function returns the rules affected and error.
+        """
+        
+        no_exists_policy = []
+        for rule in rules:
+            if not self.get_model().has_policy(sec, ptype, rule):
+                no_exists_policy.append(rule)
+
+        if should_persist:
+            try:
+                if isinstance(self.adapter, batch_adapter):
+                    self.adapter.add_policies(sec, ptype, rules)
+            except Exception as e:
+                self.logger.log("An error occurred: " + e)
+
+        self.get_model().add_policies(sec, ptype, no_exists_policy)
+
+        if sec == "g":
+            try:
+                self.build_incremental_role_links(PolicyOp.Policy_add, ptype, no_exists_policy)
+            except Exception as e:
+                self.logger.log("An exception occurred: " + e)
+                return no_exists_policy
+
+        return no_exists_policy
+
+    def remove_policy_self(self, should_persist, sec, ptype, rules):
+        """
+        remove_policy_self provides a method for dispatcher to remove policies from current policy.
+        The function returns the rules affected and error.
+        """
+        if(should_persist):
+            try:
+                if(isinstance(self.adapter, batch_adapter)):
```

Comment:
batch_adapter is a function, it should be BatchAdapter

Reply:
Also, `self.adapter` didn't exist. It should be `self._e.adapter` or `self.get_adapter()`